ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java

```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
     this.chunkSize = chunkSize;
     this.totalSize = totalSize;
     this.keys = keys;
+    long last = 0;
+    for (StoreKey key : keys) {
+      offsets.add(last);
+      keysAndSizesAndOffsets.add(new StoreKeyAndSizeAndOffset(key, last, Math.min(chunkSize, totalSize - last)));
+      last += chunkSize;
+    }
+  }
+
+  /**
+   * Construct a {@link CompositeBlobInfo} object.
+   * @param keysAndContentSizes
+   */
+  public CompositeBlobInfo(List<Pair<StoreKey,Long>> keysAndContentSizes) {
+    this.chunkSize = -1;
+    this.keys = new ArrayList<>();
+    long last = 0;
+    for (Pair<StoreKey,Long> keyAndContentSize : keysAndContentSizes) {
+      StoreKey key = keyAndContentSize.getFirst();
+      keys.add(key);
+      offsets.add(last);
+      keysAndSizesAndOffsets.add(new StoreKeyAndSizeAndOffset(key, last, keyAndContentSize.getSecond()));
+      last += keyAndContentSize.getSecond();
+    }
+    this.totalSize = last;
+  }
+
+  /**
+   * Returns the keys (along with their data content size and offset relative to the total
+   * composite blob) of the chunks related to the given byte range of the entire composite
+   * blob
+   * @param start inclusive starting byte index
+   * @param end inclusive ending byte index
+   * @return
+   */
+  public List<StoreKeyAndSizeAndOffset> getStoreKeysInByteRange(long start, long end) {
+    if (end < start || start < 0L || end >= totalSize) {
+      throw new IllegalArgumentException("Bad input parameters, start="+start+" end="+end+" totalSize="+totalSize);
+    }
+    int idx = Collections.binarySearch(offsets, start);
+    //binarySearch returns -(insertion point) - 1 if not an exact match, which points to the index of
+    //the first element greater than the key, or list.size() if all elements in the list are
+    // less than the specified key.
+    if (idx < 0) {
+      idx = -idx-2; //points to element with offset closest to but less than 'start'
+    }
+    List<StoreKeyAndSizeAndOffset> ans = new ArrayList<>();
```

Comment:
We can, and another cool thing about using `subList()` is that I can do another binary search for the end index replacing the linearly scaling while loop and make this method a O(log(n)) operation (since `subList()` itself is a constant operation for `ArrayList`).  Great suggestion!